### PR TITLE
[ROCm] use ROCm 1.9.2 fp16 capabilities in rocBLAS and MIOpen interfaces

### PR DESCRIPTION
**Pull request by iotamudelta**
* relax MIOpen if statement to allow fp16/fp32 mixed precision training now supported by ROCm 1.9.2
* use gemm_ex API of rocBLAS in ROCm 1.9.2 instead of the previous hgemm API
* with this: enable all but one half test in test_nn

While there, fix also:
* a group convolution issue w/ MIOpen pertaining to initializing MIOpen on multi-GPU systems properly we detected while working on this

### CODE REVIEW

**Comment by iotamudelta** [comment]
@bddppq @ezyang for awareness

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 1]
```diff
@@ -230,6 +230,7 @@ struct ConvolutionParams
   int dilation[max_dim];
   int64_t groups;
   bool deterministic;
+  int device_id; //This is needed to distinguish between miopen handles of multiple gpus.
```
Comment (bddppq) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp]: Which places use it?

File: aten/src/THC/THCBlas.cu [hunk 2]
```diff
@@ -286,10 +286,19 @@ void THCudaBlas_Hgemm(THCState *state, char transa, char transb, int64_t m, int6
       cublasSetStream(handle, THCState_getCurrentStream(state));
 
 #ifdef __HIP_PLATFORM_HCC__
+#if __hcc_workweek__ < 18451
```
Comment (bddppq) [inline aten/src/THC/THCBlas.cu]: I suggest let's only support single version of ROCm for now as the whole platform is moving quickly.

File: aten/src/THC/generic/THCTensorMathBlas.cu [hunk 3]
```diff
@@ -687,8 +687,10 @@ void THCTensor_(baddbmm)(THCState *state, THCTensor *result, scalar_t beta, THCT
         THCTensor_(data)(state, result_) + i * result_->stride(0), ldc);
   }
 #else
+#ifndef __HIP_PLATFORM_HCC__
   cudaDeviceProp* prop = THCState_getCurrentDeviceProperties(state);
   if (prop->major >= 5){
```
Comment (bddppq) [inline aten/src/THC/generic/THCTensorMathBlas.cu]: does hip also need to check the major version? https://github.com/pytorch/pytorch/blob/7e2b074219fad6d2b09b379423e83b2295b29df2/caffe2/core/common_gpu.h#L81-L86

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 1 above)
Comment (iotamudelta) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp]: Line 267, see below.

File: aten/src/THC/THCBlas.cu (hunk 2 above)
Comment (iotamudelta) [reply aten/src/THC/THCBlas.cu]: Fair enough, will clean this out.

File: aten/src/THC/generic/THCTensorMathBlas.cu (hunk 3 above)
Comment (iotamudelta) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: gemm_ex is supported on all ROCm supported gfx architectures, performance for mixed precision will be different based on architecture characteristics

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp]: I saw that, but it's just setting this field, I don't see anyone is reading/using it though.

File: aten/src/THC/generic/THCTensorMathBlas.cu (hunk 3 above)
Comment (iotamudelta) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: I disagree. Only for gfx900 will it not have faster intrinsics so performance will be comparable to sgemm. However, you still profit from slightly reduced memory bandwidth usage and gain accuracy over straight hgemm. From gfx906 on, it'll be fast and accurate. Debugging arch-dependent accuracy issues doesn't sound like fun to me...

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 1 above)
Comment (lcskrishna) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp]: Its used in Line 494 in the method findAlgorithm.

File: aten/src/THC/generic/THCTensorMathBlas.cu (hunk 3 above)
Comment (rohithkrn) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: @iotamudelta In Caffe2, I used major version 3 (gfx9) which I thought is practically the min. version we support.

Comment (bddppq) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: @rohithkrn @iotamudelta Could you confirm whether this is supported in all archs or only in major version >= 3?

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp]: @lcskrishna I see. Thanks for clarification.

File: aten/src/THC/generic/THCTensorMathBlas.cu (hunk 3 above)
Comment (iotamudelta) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: @bddppq rocBLAS gemm_ex is supported on all uarchs that ROCm supports (so even below gfx900 / below major version 3). What @rohithkrn is pointing out is that we never did any tests on or even compilations for < 3.

Comment (bddppq) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: Ok I see. In this case, @rohithkrn could you change caffe2 to also not check the major version? (in a separate PR)

**Review by bddppq — APPROVED** [review]

File: aten/src/THC/generic/THCTensorMathBlas.cu (hunk 3 above)
Comment (rohithkrn) [reply aten/src/THC/generic/THCTensorMathBlas.cu]: will do

**Pull request closed without merging**